### PR TITLE
add thread helper

### PR DESCRIPTION
**Pull request by lffspaniol**
This PR creates a helper package that works with errorgroups

Risks:
- Use without proper caution may result in race conditions

### CODE REVIEW

File: lib/thread.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package lib
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// ErrorGroup resolve all go routine, if some one fail will return the first error
```
Comment (brunofurmon) [inline lib/thread.go:9]:
```suggestion
// ErrorGroup resolves all go routines. It fails at first error encountered
```

File: lib/thread_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		ctx  context.Context
+		args []func() error
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "should not return an error",
```
Comment (brunofurmon) [inline lib/thread_test.go:20]:
```suggestion
			name: "Should run successfully when all go routines complete successfully",
```

File: lib/thread_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+						return nil
+					},
+					func() error {
+						return nil
+					},
+				},
+			},
+		},
+		{
+			name: "should return an error",
```
Comment (brunofurmon) [inline lib/thread_test.go:34]:
```suggestion
			name: "Should fail when first fo routine fails",
```

File: lib/thread_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			args: args{
+				ctx: context.Background(),
+				args: []func() error{
+					func() error {
+						return errors.New("ops, you recive an error")
```
Comment (brunofurmon) [inline lib/thread_test.go:39]:
```suggestion
						return errors.New("Oops, you received an error")
```

**Review by brunofurmon — APPROVED** [review]

**Review by Scherpel — APPROVED** [review]

File: lib/thread.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	g, _ := errgroup.WithContext(ctx)
+
+	for _, fn := range args {
+		g.Go(fn)
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+	return nil
```
Comment (bernardolm) [inline lib/thread.go:20]:
```suggestion
	}
	
	return nil
```

File: lib/thread.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+	"golang.org/x/sync/errgroup"
+)
+
+// ErrorGroup resolves all go routines. It fails at first error encountered
+func ErrorGroup(ctx context.Context, args ...func() error) error {
+	g, _ := errgroup.WithContext(ctx)
+
+	for _, fn := range args {
+		g.Go(fn)
```
Comment (bernardolm) [inline lib/thread.go:14]: go routine?

Comment (brunofurmon) [reply lib/thread.go:14]: Seems more like a `Parallel.Invoke`, this method could even be called with itself as an argument...
Unless there is a way good way of separating any of the arg funcs as a goroutine or simple call, one way would be passing funcs as an array of an interface that explicitly identifies it

**Pull request merged**